# Proyecto RAG con LLM y métricas de evaluación

Este notebook implementa un **sistema RAG (Retrieval-Augmented Generation)** que combina:

1. **Carga y procesamiento de documentos** (PDFs reales).
2. **Chunking y generación de embeddings** para cada fragmento.
3. **Búsqueda semántica con FAISS HNSW** y BM25 híbrido.
4. **Reranking con Cross-Encoder** para mejorar precisión.
5. **Generación de respuestas con un LLM (Qwen 4-bit)**.
6. **Validación y métricas**: grounding score, BLEU, ROUGE, y citas por oración.

El objetivo es **responder preguntas usando solo la información contenida en los documentos**, minimizando alucinaciones y asegurando trazabilidad de las respuestas.

### 0. Instalación de librerías
Se instalan todas las dependencias necesarias:
- `transformers`, `accelerate`, `bitsandbytes`, `sentencepiece` → para LLM y cuantización 4-bit.
- `faiss-cpu`, `sentence-transformers`, `datasets` → para embeddings y FAISS.
- `rank_bm25`, `rouge_score`, `nltk`, `pypdf` → para recuperación híbrida, evaluación automática y manejo de PDFs.

Garantiza que todo el pipeline (RAG, reranking, métricas, LLM) pueda ejecutarse sin errores por falta de librerías.

### Instalación de librerías

In [1]:
!pip install -U transformers accelerate bitsandbytes sentencepiece
!pip install faiss-cpu sentence-transformers datasets
!pip install rank_bm25 rouge_score nltk pypdf

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.7/10.7 MB 85.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 13.1 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 5.0.0
    Uninstalling transformers-5.0.0:
      Successfully uninstalled transformers-5.0.0
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.8/23.8 MB 73.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 333.7/333.7 kB 28.5 MB/s eta 0:00:00
  Created wheel for rouge_score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=6099d76e9595321e96d1503f37d7d6ecff0f44360b2b345ae3f0d98fb60c636c
  Stored in directory: /root/.cache/pip/wheels/85/9d/af/01feefbe7d55ef5468796f0c68225b6788e85d9d0a281e7a70
Successfully built rouge_score


### 1. Preparación de carpeta de documentos
Se crea la carpeta `/content/docs` donde se subirán los PDFs que formarán el dataset. Esto asegura que el pipeline RAG tenga un lugar definido para leer los documentos.

In [4]:
# comandos para crear un carpeta en los contenidos de nombre "docs", donde almacenaremos los documentos
import os
os.makedirs("/content/docs", exist_ok=True)
print("Carpeta lista — sube tus PDFs a la carpeta docs en el panel de archivos")

Carpeta lista — sube tus PDFs a la carpeta docs en el panel de archivos


### 2. Dataset (Carga de PDFs)
Se leen los PDFs de la carpeta, página por página, limpiando saltos de línea y espacios. Cada página se guarda en `texts` y se le asigna un título en `titles` Esto prepara los documentos para el pipeline RAG. Se imprime la cantidad de páginas cargadas y un ejemplo de contenido.

In [5]:
from pypdf import PdfReader
import re, os

# creamos dos listas vacias, para almacenar la informacion
texts, titles = [], []

# Listamos todos los archivos dentro de la carpeta /content/docs.
for filename in os.listdir("/content/docs"):
    # Verifica que el archivo termine en .pdf. Si no es PDF, lo salta (continue)
    if not filename.endswith(".pdf"):
        continue
    # Carga el PDF para poder leer sus páginas.
    reader   = PdfReader(f"/content/docs/{filename}")
    # Obtienemos el nombre del archivo sin la extensión .pdf.
    doc_name = filename.replace(".pdf", "")

    # Recorrer páginas del PDF
    # Extraemos el texto de la página, si no hay texto (None) = cadena vacía.
    # eliminamos saltos de línea "\n", reemplazamos espacios por uno solo.
    # strip(): elimina espacios al inicio y final.
    for i, page in enumerate(reader.pages):
        texto = page.extract_text() or ""
        texto = re.sub(r"\s+", " ", texto.replace("\n", " ")).strip()
        if len(texto) > 100:
            texts.append(texto)
            titles.append(f"{doc_name} · p{i+1}")
    #-----------------------------------------------------------

# imprimimos cuantas paginas se cargaron
print("Páginas cargadas:", len(texts))
print("Ejemplo —", titles[0], ":", texts[0][:200])

Páginas cargadas: 52
Ejemplo — 1660-2025.R · p1 : Lima, 09 de mayo de 2025 « Resolución S.B.S.» N° 1660-2025 El Superintendente de Banca, Seguros y Administradoras Privadas de Fondos de Pensiones CONSIDERANDO: Que, el numeral 13 del artículo 349 de l


### 3. Chunking con Overlap

En esta sección se divide el texto en fragmentos (chunks) con solapamiento(overlap) para mantener continuidad entre ellos.

**Tradeoff (equilibrio) del tamaño de chunk**
- Chunk pequeño → búsqueda precisa, pero puede perder contexto
- Chunk grande → más contexto para el LLM, pero menos precisión al buscar

**El Tradeoff (equilibrio) del overlap:**
- Chunk pequeño → búsqueda precisa, pero puede perder contexto
- Chunk grande → más contexto para el LLM, pero menos precisión al buscar

Se utiliza `size=400, overlap=100`  como un equilibrio entre contexto, precisión y eficiencia.

In [6]:
# Defines una función llamada chunk_text con parametros (text, title, size=400, overlap=100):
# size = tamaño de 400 caracteres
# overlap = solapamiento de 100 caracteres
def chunk_text(text, title, size=400, overlap=100):
    # Lusta vacia para almacenamiento
    chunks = []
    # fijamos un valor (dónde empezar a cortar el texto.)
    start = 0
    # bucle que se ejecuta mientras no haya recorrido toda la longitud de texto
    while start < len(text):
        # Agregamos un nuevo elemento a la lsita chunk y respectivo titulo
        chunks.append({"text": text[start:start+size], "title": title})
        start += size - overlap # <---- aquí se crea el "overlap": cada chunk comparte parte del texto con el anterior
    return chunks
# ------------------------------------------------------------------------------

# definimos una lista vacia
all_chunks = []
# recursivamente agregamos los chunks y sus titulos definidos en la lista all_chunks
for text, title in zip(texts, titles):
    all_chunks.extend(chunk_text(text, title))

# listamos los pedazos de texto con los títulos correspondientes.
chunk_texts  = [c["text"]  for c in all_chunks]
chunk_titles = [c["title"] for c in all_chunks]

# cantidad de chunk generado
print("Total chunks:", len(all_chunks))

Total chunks: 541


### 4. Embeddings

Se convierten los chunks de texto en vectores numéricos (embeddings) usando un modelo pre-entrenado. Cada vector representa el significado de un chunk, lo que permite comparar y buscar información semántica de manera eficiente. (Textos con significado similar quedan cerca en el espacio vectorial).

In [7]:
# librería para generar embeddings a partir de texto.
# numpy (np): para manipular vectores/matrices de manera eficiente.
from sentence_transformers import SentenceTransformer
import numpy as np

# Cargamos un modelo pre-entrenado.
embedder   = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")

# Convertimos cada texto de "chunk_texts" en un vector numérico (embedding).
embeddings = embedder.encode(chunk_texts, show_progress_bar=True)
# convertimos la lista de vectores "embeddings" en un array de Numpy.
embeddings = np.array(embeddings).astype("float32")

# mostramos de que dimension es la matriz generada
print("Shape:", embeddings.shape)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/645 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/526 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Batches:   0%|          | 0/17 [00:00<?, ?it/s]

Shape: (541, 384)


### 5. FAISS HNSW
Se utiliza **FAISS** con **HNSW** (Hierarchical Navigable Small World) para indexar los embeddings de los chunks y permitir búsquedas semánticas rápidas.

**HNSW** organiza los vectores en capas jerárquicas:
- Navega rápido por capas altas y refina en capas bajas, como un GPS.
- Ventaja sobre IVFFlat: no requiere entrenamiento previo y ofrece mejor recall.

**Parámetros importantes**:
- `M=32` → conexiones por nodo (más = mejor calidad, más memoria)
- `efConstruction=200` → esfuerzo al construir
- `efSearch=50` → esfuerzo al buscar vecinos

In [8]:
# FAISS es una librería de Facebook AI para búsqueda rápida de vectores.
# útil para encontrar chunks semánticamente similares en grandes colecciones.
import faiss

# número de columnas de tu array de embeddings
dim = embeddings.shape[1]
# Convierte cada vector a longitud 1 (vector unitario) usando la norma L2
faiss.normalize_L2(embeddings)

#  creamos el índice hnsw basado en HNSW (Hierarchical Navigable Small World)
# con dimension generada en la linea superior y número de vecinos M en el grafo HNSW (afecta precisión/velocidad)
hnsw = faiss.IndexHNSWFlat(dim, 32)

# controlamos precisión al construir el grafo
hnsw.hnsw.efConstruction = 200
# controlamos precisión en la búsqueda
hnsw.hnsw.efSearch = 50
# Inserta todos los embeddings (4964 vectores) en el índice HNSW
hnsw.add(embeddings)

# imprimos número total de vectores agregados al índice
print("Vectores indexados:", hnsw.ntotal)

Vectores indexados: 541


### 6. Búsqueda Híbrida: BM25 + FAISS

Para mejorar la recuperación de información se combina:

-**BM25** = búsqueda clásica por palabras clave.

-**FAISS** = búsqueda semántica basada en embeddings (por significado).

La combinación se hace mediante una mezcla lineal controlada por un parámetro alpha:

**score = alpha × FAISS + (1 - alpha) × BM25**

**FAISS** y **BM25** se normalizan entre 0 y 1 para que los scores sean comparables.

- alpha decide la importancia relativa de cada método:

    (i) alpha cercano a 1 → más peso a FAISS (semántica)

    (ii) alpha cercano a 0 → más peso a BM25 (palabras clave)

- Ejemplo típico: score = 0.7 × FAISS + 0.3 × BM25

**Ventajas:**

- Cubre las limitaciones de cada método: FAISS maneja sinónimos y contexto BM25 garantiza coincidencias literales precisas.

- **La función retrieve()** devuelve los k chunks más relevantes según la consulta, integrando semántica y palabras clave de manera eficiente.

In [9]:
# importando la clase BM25Okapi de la librería rank_bm25
# algoritmo de recuperación de información basado en palabras clave.
from rank_bm25 import BM25Okapi

# Conviertimos cada chunk a lista de palabras minúsculas.
tokenized = [t.lower().split() for t in chunk_texts]
# Inicializamos BM25 con todos los chunks tokenizados (búsquedas por palabras clave)
bm25      = BM25Okapi(tokenized)

# definimos función llamada ** retrieve ** el cual sirve para recuperar los textos más relevantes
# query = la frase o palabra a buscar
# k = número de resultados que quieres devolver
# alpha = peso que mezcla FAISS (semántica) y BM25 (palabras clave)
def retrieve(query, k=10, alpha=0.7):
    # listamos de todos los pedazos de texto
    n = len(chunk_texts)

    # FAISS
    # Convertimos la consulta en un embedding vectorial, Normaliza (similitud coseno.)
    # y Busca en el índice HNSW los k*3 vecinos más cercanos
    q_emb = embedder.encode([query]).astype("float32")
    faiss.normalize_L2(q_emb)
    D, I  = hnsw.search(q_emb, min(k*3, n))
    # -------------------------------------

    # Crea un array con ceros para todos los chunks y Asignamos la similitud FAISS solo a los chunks encontrados.
    # por el bucle defindo
    faiss_scores = np.zeros(n)
    for s, i in zip(D[0], I[0]):
        if i >= 0: faiss_scores[i] = float(s)
        # -----------------------------------

    # Escala los scores entre 0 y 1 para que sean comparables con BM25.
    if faiss_scores.max() > 0:
        faiss_scores /= faiss_scores.max()

    # BM25
    # devuelve un score de relevancia para cada chunk según la consulta.
    bm25_raw = bm25.get_scores(query.lower().split())
    # Normaliza BM25 también entre 0 y 1.
    bm25_scores = bm25_raw / bm25_raw.max() if bm25_raw.max() > 0 else bm25_raw

    # Fusión
    # Mezclamos resultados de FAIS y BM25 según un peso alpha.
    # Ordenamos de mayor a menor los chunks más relevantes
    # Se toman solo los k mejores resultado
    hybrid  = alpha * faiss_scores + (1 - alpha) * bm25_scores
    top_idx = np.argsort(hybrid)[::-1][:k]

    # Devuelve una lista: text =el chunk y title = su título original
    return [{"text": chunk_texts[i], "title": chunk_titles[i]} for i in top_idx]

### 7. Reranker Cross-Encoder

FAISS usa un **bi-encoder**: codifica pregunta y documento por separado, luego compara vectores. Rápido pero aproximado.

El reranker usa un **cross-encoder**: analiza pregunta + documento juntos. Mucho más preciso.

Estrategia del pipeline:
1. FAISS recupera top-10 chunks rápido (búsqueda semántica aproximada).
2. El reranker reordena esos top-10 usando cross-encoder para evaluar relevancia exacta.
3. Se seleccionan los top-3 resultados para alimentar al LLM.

**Función rerank():**
- Recibe una consulta y una lista de chunks candidatos.
- Predice un score de relevancia para cada par (consulta, chunk).
- Devuelve los top-k candidatos ordenados por relevancia.

In [10]:
# Importamos librería de Python diseñada para trabajar con modelos de embeddings de oraciones o textos.
# es un modelo que evalúa la relevancia de un par de textos: (consulta, documento)
from sentence_transformers import CrossEncoder

# Cargamos un modelo pre-entrenado para reranking tipo MS MARCO.
reranker = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")

# definimos la funcion que búsca la consulta del usuario y que tenga una lista de resultados previos (chunks)
# devuelve solo los 3 mejores resultados después de reranking
def rerank(query, candidates, top_k=3):
    # Convierte cada candidato en un par (query, texto) y devuelve un score de relevancia
    scores = reranker.predict([(query, c["text"]) for c in candidates])
    # Agrega un nuevo campo "rerank_score" a cada candidato y guarda el score predicho por el CrossEncoder
    for c, s in zip(candidates, scores):
        c["rerank_score"] = float(s)
    # Retornamos ordenando de mayor a menor relevancia
    return sorted(candidates, key=lambda x: x["rerank_score"], reverse=True)[:top_k]

config.json:   0%|          | 0.00/794 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/105 [00:00<?, ?it/s]

BertForSequenceClassification LOAD REPORT from: cross-encoder/ms-marco-MiniLM-L-6-v2
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

### 8. LLM Microsoft Phi-2
El LLM es el modelo final que genera la respuesta basada en los chunks más relevantes.

**Características:**
- Phi-2 (~2.7B parámetros) de Microsoft, cargado con cuantización de 4 bits. Modelo compacto con buen razonamiento en inglés y español.
- El prompt le indica que responda **solo con el contexto** proporcionado. Esto es la primera defensa contra alucinaciones.
- A diferencia de Qwen, Phi-2 usa un formato de prompt simple (Instruct) sin chat template.

**Componentes:**
- torch → backend PyTorch para operaciones tensoriales y GPU.
- AutoTokenizer → convierte texto a tokens que el modelo entiende.
- AutoModelForCausalLM → modelo de lenguaje causal capaz de generar texto.
- BitsAndBytesConfig → configuración para carga en 4 bits y ahorro de memoria.

**Flujo en el pipeline:**
1. Recibe los top-k chunks seleccionados por FAISS + Reranker.
2. Genera la respuesta textual final respetando el contexto proporcionado.

In [11]:
# torch → PyTorch, backend para el modelo.
# AutoTokenizer → convierte texto a tokens que entiende el modelo.
# AutoModelForCausalLM → modelo causal language model, capaz de generar texto.
# BitsAndBytesConfig → configuración para cuantización (usar menos memoria).

import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

# Indica el modelo a cargar: Microsoft Phi-2 con ~2.7B parámetros
model_name = "microsoft/phi-2"

# Permite cargar el modelo en 4 bits en lugar de 16 o 32. Menos uso de GPU/CPU y acelera inferencia
bnb       = BitsAndBytesConfig(load_in_4bit=True)
# convierte texto ↔ tokens
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
# Phi-2 no tiene pad_token por defecto, se asigna el eos_token
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token
# carga los pesos del modelo desde Hugging Face, con Parámetros claves.
model     = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb,
    device_map="auto",
    trust_remote_code=True
)
# Coloca el modelo en modo inferencia, desactivando el entrenamiento
model.eval()
print("✅ Phi-2 cargado")

config.json:   0%|          | 0.00/735 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/453 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

✅ Phi-2 cargado


### 9. Prompt + generación de respuestas
Este paso utiliza el LLM (Phi-2 4-bit) para generar la respuesta final basada en los chunks más relevantes obtenidos del pipeline RAG.

Flujo:
1. Recibe la pregunta del usuario y el contexto (chunks top después de retrieval + reranker).
2. Construye un prompt en formato **Instruct** (específico de Phi-2): `Instruct: ... Output:`.
3. Convierte el prompt a tokens usando el tokenizer.
4. Genera texto con el modelo Phi-2 (quantized 4-bit para eficiencia).
5. Extrae solo la parte generada (descarta el prompt de entrada).
6. Devuelve la respuesta textual final.

**Diferencia clave con Qwen:** Phi-2 no soporta `apply_chat_template`, por lo que se usa un prompt de texto plano con el formato `Instruct / Output` que el modelo fue entrenado a seguir.

Importancia:
- Garantiza que el LLM solo use información recuperada (previene alucinaciones).
- Completa la etapa de generación del pipeline RAG.

In [12]:
# Define una función llamada ask_llm que recibe:
# 1.- context: el texto (chunks) que el LLM puede usar para responder.
# 2.- question: la pregunta que queremos hacerle al LLM.

def ask_llm(context, question):
    # Phi-2 usa formato Instruct/Output en lugar de chat template
    # Se construye el prompt directamente como texto plano
    prompt = (
        f"Instruct: Responde ÚNICAMENTE usando el contexto proporcionado. "
        f"Si la respuesta no está en el contexto, di 'No encontrado en el contexto'.\n"
        f"Contexto:\n{context}\n\n"
        f"Pregunta: {question}\n"
        f"Output:"
    )
    # Convierte el prompt en tokens que el modelo entiende y devuelve tensores de PyTorch.
    inputs = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=2048).to(model.device)
    with torch.inference_mode():
        # Genera la respuesta del LLM:
        out = model.generate(
            **inputs,
            max_new_tokens=200,
            temperature=0.3,
            do_sample=True,
            pad_token_id=tokenizer.eos_token_id
        )
    # Extrae solo los tokens generados, eliminando los tokens del prompt.
    new_tokens = out[0][inputs["input_ids"].shape[1]:]
    # Devuelve la respuesta final del LLM como string.
    return tokenizer.decode(new_tokens, skip_special_tokens=True).strip()


### 10. Citation per sentence

Este paso mejora la trazabilidad de la respuesta del LLM, agregando referencias a los chunks originales que respaldan cada oración de la respuesta.

**Flujo:**
1. Divide la respuesta generada en oraciones.

2. Para cada oración, calcula el "overlap" de palabras con cada chunk.

3. Asigna como cita el título del chunk con mayor overlap.

4. Si no hay coincidencia significativa (>10% palabras), marca "no encontrado".

5. Devuelve la respuesta con citas por oración.

**Importancia:**
Permite validar la fuente de cada afirmación.
- Aumenta transparencia y confiabilidad del sistema RAG.
- Útil en contextos académicos o de documentación.

In [13]:
# Importa la librería re de Python para trabajar con expresiones regulares, usada aquí para dividir texto en oraciones.
import re

# Define una función llamada cite_sentences que recibe:
# answer: la respuesta generada por el LLM.
# chunks: la lista de chunks originales (cada uno con text y title) para buscar coincidencias.
def cite_sentences(answer, chunks):
    # Divide la respuesta en oraciones usando regex:
    sentences = re.split(r'(?<=[.!?])\s+', answer.strip())
    # Filtra oraciones demasiado cortas (menos de 10 caracteres).
    sentences = [s for s in sentences if len(s) > 10]
    # Inicializa una lista vacía donde se guardarán las oraciones con su cita.
    result = []

    # Recorre cada oracion en minúscula,
    # cada título del chunk con mayor coincidencia (inicialmente "no encontrado")
    # y  puntaje de overlap máximo. (inicialmente -1).
    for sent in sentences:
        words = set(sent.lower().split())
        best, best_score = "no encontrado", -1
        # Convierte el texto del chunk a palabras en minúscula,
        # calcula el overlap y si es mayor que best_score, actualiza best_score y best con el título de ese chunk.
        for c in chunks:
            overlap = len(words & set(c["text"].lower().split())) / max(len(words), 1)
            if overlap > best_score:
                best_score, best = overlap, c["title"]
        # Solo considera que un chunk respalda la oración si más del 10% de las palabras coinciden (best_score > 0.1).
        # Si no, marca "no encontrado".
        citation = best if best_score > 0.1 else "no encontrado"
        result.append(f"{sent} [{citation}]")
        # ---------------------------------------------------
    # Devuelve todas las oraciones con citas como un solo string
    return "\n".join(result)

### 11. Hallucination guard + grounding score

Este paso verifica que la respuesta del LLM esté respaldada por el contexto.
Flujo:
1. Extrae todas las palabras de los chunks recuperados.
2. Filtra palabras irrelevantes (stopwords y cortas) de la respuesta.
3. Calcula el grounding score: proporción de palabras de la respuesta presentes en el contexto.
4. Asigna un veredicto según el score:
- ≥0.7 ✅ FUNDAMENTADO
- 0.4–0.7 ⚠️ ADVERTENCIA
- <0.4 ❌ ALUCINACIÓN
5. Devuelve el score y el veredicto.
Importancia:
- Evita respuestas inventadas por el modelo (alucinaciones).
- Garantiza que la información provenga del contenido recuperado.

In [14]:
# Define la función hallucination_guard que recibe dos parámetros:
# answer: la respuesta generada por el LLM.
# chunks: lista de diccionarios con los fragmentos de texto originales recuperados
def hallucination_guard(answer, chunks):
    # Junta todo el texto de los chunks en un solo string
    ctx_words = set(" ".join(c["text"] for c in chunks).lower().split())
    # Definimos palabras comunes (stopwords) que no aportan significado importante, para ignorarlas en la comparación.
    stopwords = {"el","la","los","las","un","una","es","son","en","de","y","a","the","is","of","and","to","a"}
    # Toma la respuesta answer, la pasa a minúsculas y la divide en palabras.
    content   = [w for w in answer.lower().split() if w not in stopwords and len(w) > 2]
    # filtrado no queda ninguna palabra significativa (content vacío)
    if not content:
        return 0.0, "SIN DATOS"
    # Calcula cuántas palabras relevantes de la respuesta aparecen en el contexto (ctx_words).
    score   = sum(1 for w in content if w in ctx_words) / len(content)
    # Asigna un veredicto según el grounding score
    verdict = "✅ FUNDAMENTADO" if score >= 0.7 else ("⚠️ ADVERTENCIA" if score >= 0.4 else "❌ ALUCINACIÓN")
    # Devuelve el score redondeado a 4 decimales y el veredicto correspondiente.
    return round(score, 4), verdict

### 12. BLEU + ROUGE
Este paso permite medir cuantitativamente la calidad de la respuesta generada comparándola con una referencia “correcta”.

Flujo:
1. Tokeniza y normaliza a minúsculas la respuesta generada y la referencia.
2. Calcula BLEU-1 → mide coincidencia de palabras individuales (unigrams).
3. Calcula ROUGE-L → mide coincidencia de secuencias largas (longest common subsequence).
4. Devuelve un diccionario con BLEU-1 y ROUGE-L redondeados a 4 decimales.

**Importancia:**
- Evalúa la fidelidad de la generación automática.
- Permite comparar distintas estrategias de retrieval, prompts o reranking.
- Complementa el grounding score para medir precisión y coherencia.

In [15]:
# Importa librerías necesarias:
# nltk → procesamiento de texto y tokenización.
# sentence_bleu → calcula la métrica BLEU (coincidencia de palabras n-gram) a nivel de oración.
# SmoothingFunction → evita problemas cuando no hay coincidencias exactas.
# rouge_scorer → calcula la métrica ROUGE, evaluando similitud entre resumen o texto generado y referencia.

import nltk
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge_score import rouge_scorer


# Descargamos los modelos de tokenización de NLTK.
# para separar frases y palabras (quiet=True es decir no muestra logs de descarga).
nltk.download("punkt",     quiet=True)
nltk.download("punkt_tab", quiet=True)

# Inicializa el scorer de ROUGE:
# "rouge1" → compara coincidencia de unigrams (palabras individuales).
# "rougeL" → compara la longest common subsequence (frases o secuencia de palabras).
rouge    = rouge_scorer.RougeScorer(["rouge1", "rougeL"], use_stemmer=True)
# función para suavizar scores BLEU cuando hay pocas coincidencias, evitando 0 exacto.
smoother = SmoothingFunction().method1

# Definimos la función evaluate que recibe:
# generated → respuesta del LLM.
# reference → respuesta “correcta” o esperada para comparar.
def evaluate(generated, reference):
    ref   = reference.lower().split()
    gen   = generated.lower().split()
    # Calcula BLEU-1: proporción de unigrams de generated que coinciden con reference.
    bleu1 = sentence_bleu([ref], gen, weights=(1,0,0,0), smoothing_function=smoother)
    # Calcula métricas ROUGE entre referencia y texto generado.
    rs    = rouge.score(reference, generated)
    # Devuelve un diccionario con métricas evaluadas:
    return {
        "BLEU-1":  round(bleu1, 4), # proporción de palabras coincidentes.
        "ROUGE-L": round(rs["rougeL"].fmeasure, 4), # proporción de secuencias largas coincidentes.
    }

### 13. Consulta
Aquí el usuario hace la pregunta y se ejecuta todo el pipeline: retrieve → rerank → LLM → hallucination guard.

(i) Se obtiene el contexto de los chunks más relevantes.

(ii) El LLM genera la respuesta usando solo ese contexto.

(iii) Se calcula el grounding score para medir si la respuesta está fundamentada.

**Objetivo:** devolver una respuesta confiable basada en los documentos originales.

In [21]:
query = "¿Qué es Gestión de activos y pasivos?"  # ← cambia por tu pregunta (Pregunta del usuario)

# Recuperación: se buscan los chunks más relevantes (FAISS + BM25)
docs    = retrieve(query, k=10)
# Reranking: top-10 se reordena con Cross-Encoder, quedando top-3
top     = rerank(query, docs, top_k=3)
# Contexto: se unen los textos de los chunks rerankeados
context = "\n".join(c["text"] for c in top)
# Generación: el LLM Qwen produce la respuesta usando solo este contexto
answer  = ask_llm(context, query)
# Grounding: se evalúa si la respuesta está fundamentada en los chunks
score, verdict = hallucination_guard(answer, top)

# Salida: se imprime la respuesta y su grounding score
print("RESPUESTA:\n", answer)
print("\nGrounding score:", score, verdict)

RESPUESTA:
 Gestión de activos y pasivos es la gestión de riesgos y desventajas que afectan a la implementación de los procesos de gestión de activos y pasivos en empresas de seguros.

Grounding score: 0.7333 ✅ FUNDAMENTADO


**concluimos lo siguiente:**

En esta consulta sobre la Gestión de Activos y Pasivos, el sistema arrojó un Grounding Score de 0.73. Desde una perspectiva analítica, este valor representa un coeficiente de fidelidad semántica: certifica que el 73% de los términos técnicos y conceptos clave de la respuesta fueron extraídos directamente de nuestra base de datos normativa indexada

El veredicto FUNDAMENTADO nos da la seguridad de que el modelo Phi-2 ha procesado correctamente los fragmentos recuperados, entregando una definición técnica precisa y libre de alucinaciones. Con esto, demostramos que el pipeline no solo recupera información, sino que garantiza que la generación final sea veraz y auditable.

### Varias consultas

In [17]:
questions = [
    ("¿Qué flujos se consideran en instrumentos de deuda para el ASA?",
     "En instrumentos de deuda se consideran los intereses y amortizaciones en las fechas correspondientes."),

    ("¿Cuál es el nivel mínimo del indicador de cobertura de los GHO regulatorios?",
     "Las empresas deben monitorear que los GHO regulatorios se encuentren respaldados por activos, manteniendo un indicador de cobertura por encima de 1.0."),

    ("¿Cuál es una función del Comité de Gestión de Activos y Pasivos?",
     "El Comité de GAP debe implementar las estrategias, planes, herramientas, planes de contingencia y otros recursos empleados para la gestión de activos y pasivos.")
]

results = []

for query, reference in questions:
    docs    = retrieve(query, k=10)
    top     = rerank(query, docs, top_k=3)
    context = "\n".join(c["text"] for c in top)
    answer  = ask_llm(context, query)
    cited   = cite_sentences(answer, top)
    score, verdict = hallucination_guard(answer, top)
    metrics = {"grounding": score, "verdict": verdict}
    metrics.update(evaluate(answer, reference))
    results.append({"query": query, "answer": answer, "cited": cited, "metrics": metrics})

    print("="*60)
    print("Q:", query)
    print("\nRespuesta con citas:")
    print(cited)
    print("\nMétricas:", metrics)

Q: ¿Qué flujos se consideran en instrumentos de deuda para el ASA?

Respuesta con citas:
En el caso de instrumentos de deuda con opción de prepago de precio fijo se incorporan todos los flujos hasta antes de la fecha de activación de la opción y, en dicha fecha, se incorpora el saldo por amortizar, más los intereses y comisiones contemplados en el evento de prepago del instrumento. [1660-2025.R · p25]
Se deben consid [1660-2025.R · p25]

Métricas: {'grounding': 1.0, 'verdict': '✅ FUNDAMENTADO', 'BLEU-1': 0.1636, 'ROUGE-L': 0.2466}
Q: ¿Cuál es el nivel mínimo del indicador de cobertura de los GHO regulatorios?

Respuesta con citas:
Responde ÚNICAMENTE usando el contexto proporcionado. [1660-2025.R · p19]
Si la respuesta no está en el contexto, di 'No encontrado en el contexto'. [1660-2025.R · p8]
Contexto:
e de la gestión de activos y pasivos a nivel del sistema de seguros, las empresas deben agrupar sus obligaciones en los siguientes GHO regulatorios: cinco (5) para ramos generales, nu

### Reporte final

In [22]:
print(f"{'='*65}")
print(f"{'Pregunta':<30} {'Grounding':>10} {'BLEU-1':>8} {'ROUGE-L':>8}  Veredicto")
print("-"*65)
for r in results:
    q  = r['query'][:28] + ".." if len(r['query']) > 28 else r['query']
    g  = r['metrics']['grounding']
    b1 = r['metrics'].get('BLEU-1', '-')
    rl = r['metrics'].get('ROUGE-L', '-')
    v  = r['metrics']['verdict']
    print(f"{q:<30} {g:>10.4f} {str(b1):>8} {str(rl):>8}  {v}")
print("="*65)

Pregunta                        Grounding   BLEU-1  ROUGE-L  Veredicto
-----------------------------------------------------------------
¿Qué flujos se consideran en..     1.0000   0.1636   0.2466  ✅ FUNDAMENTADO
¿Cuál es el nivel mínimo del..     0.7586   0.0989   0.1525  ✅ FUNDAMENTADO
¿Cuál es una función del Com..     0.8750   0.2917   0.3077  ✅ FUNDAMENTADO


### Cuadro de Métricas de Evaluación

* **Grounding Score (Fidelidad al Contexto):** Es nuestra Métrica Crítica. Un score > 0.70 (como los obtenidos en nuestras pruebas) asegura que la información es veraz y proviene exclusivamente de la Resolución SBS 1660-2025, eliminando cualquier riesgo de alucinación por parte del modelo.

* **BLEU-1 (Precisión Léxica):** Evalúa la coincidencia exacta de palabras entre la IA y la referencia humana. Valores entre 0.1 y 0.3 son aceptables en este proyecto, ya que el modelo Phi-2 genera respuestas técnicas más extensas y detalladas que el resumen de referencia.

* **ROUGE-L (Coherencia Estructural):** Mide la secuencia lógica más larga de palabras comunes. Esta métrica garantiza que la respuesta no sea solo una lista de términos sueltos, sino una explicación fluida y gramaticalmente correcta según la estructura de la normativa.



1️⃣ Consulta: “¿Qué flujos se consideran en instrumentos de deuda para el ASA?”

**Respuesta con citas:** El modelo extrajo datos precisos sobre intereses, amortizaciones y opciones de prepago directamente de la página 25 del documento.

**Grounding score:** 1.0000 → ✅ FUNDAMENTADO. Coincidencia total. El 100% de los términos técnicos provienen del contexto recuperado. Es una respuesta de máxima fidelidad.

**Interpretación:** El sistema es capaz de realizar una extracción quirúrgica de datos financieros complejos sin añadir información externa.

2️⃣ Consulta: “¿Cuál es el nivel mínimo del indicador de cobertura de los GHO regulatorios?”

**Respuesta con citas:** Se identificaron correctamente los Grupos Homogéneos de Obligaciones (GHO) citando las páginas 8 y 19.

**Grounding score:** 0.7586 → ✅ FUNDAMENTADO. Un sólido 76% de la respuesta está anclado en la normativa.

**Interpretación:** Aunque la estructura lingüística varía respecto a la fuente, el núcleo técnico sobre la cobertura de activos es veraz y está plenamente respaldado.

3️⃣ Consulta: “¿Cuál es una función del Comité de Gestión de Activos y Pasivos?”

**Respuesta con citas:** Citado correctamente desde la página 5, detallando la periodicidad de las sesiones y canales de comunicación.

**Grounding score:** 0.8750 → ✅ FUNDAMENTADO. Casi el 90% de fidelidad semántica.

**Interpretación:** El modelo demuestra una alta comprensión de las funciones administrativas y de gobernanza descritas en el manual de gestión de riesgos.